builtin/providers/cloudstack/resource_cloudstack_security_group.go

```diff
@@ -0,0 +1,210 @@
+package cloudstack
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xanzy/go-cloudstack/cloudstack"
+)
+
+func resourceCloudStackSecurityGroup() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceCloudStackSecurityGroupCreate,
+		Read:   resourceCloudStackSecurityGroupRead,
+		Delete: resourceCloudStackSecurityGroupDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+				ForceNew: true,
+			},
+
+			"project": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"ingress_rules": &schema.Schema{
+				Type:     schema.TypeList,
+				ForceNew: true,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"cidr": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+
+						"protocol": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+
+						"start_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+							ForceNew: true,
+						},
+
+						"end_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceCloudStackSecurityGroupCreate(d *schema.ResourceData, meta interface{}) error {
+	cs := meta.(*cloudstack.CloudStackClient)
+
+	name := d.Get("name").(string)
+
+	// Create a new parameter struct
+	p := cs.SecurityGroup.NewCreateSecurityGroupParams(name)
+
+	// Set the description
+	if description, ok := d.GetOk("description"); ok {
+		p.SetDescription(description.(string))
+	} else {
+		p.SetDescription(name)
+	}
+
+	// If there is a project supplied, we retrieve and set the project id
+	if err := setProjectid(p, cs, d); err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Creating security group %s", name)
+	r, err := cs.SecurityGroup.CreateSecurityGroup(p)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("[DEBUG] Security group %s successfully created with ID: %s", name, r.Id)
+	d.SetId(r.Id)
+
+	// Create Rules
+	ingressRules := d.Get("ingress_rules").([]interface{})
+	log.Printf("[DEBUG] rules=%#v\n", ingressRules)
+	for _, ingressRule := range ingressRules {
+		log.Printf("[DEBUG] rule=%#v\n", ingressRule)
+
+		rule := ingressRule.(map[string]interface{})
+
+		param := cs.SecurityGroup.NewAuthorizeSecurityGroupIngressParams()
+
+		param.SetSecuritygroupid(d.Id())
+
+		if cidr, ok := rule["cidr"]; ok {
```

Comment:
You don't have to run all these tests (`if ... ; ok`) here as they are required parameters. So they are guaranteed to be here.